From bde15b4411da48d3ecf7e131ed3bb6fe911947d4 Mon Sep 17 00:00:00 2001
From: Girish Kumar M P <quic_gmp@quicinc.com>
Date: Thu, 18 Jul 2024 23:28:55 +0530
Subject: [PATCH] wifi: ath12k: Fix pcic write crash during reboot

Currently, ab->mem iounmap and become NULL inside
ath12k_qmi_free_target_mem_chunk() for hybrid bus.
When we run reboot/rmmod then ath12k_pcic_ipci_write32()
gets invalid address (because ab->mem is NULL) during
hardware group stop and leads to crash.

Signed-off-by: Girish Kumar M P <quic_gmp@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 23 ++++++++++++-----------
 1 file changed, 12 insertions(+), 11 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index aa2a8e0..0bb2748 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -3992,18 +3992,19 @@ static void ath12k_core_hw_group_stop(struct ath12k_hw_group *ag)
 		if (!dp->reoq_lut.vaddr)
 			break;
 
-		if (dp->reoq_lut.vaddr) {
-			address = HAL_SEQ_WCSS_UMAC_REO_REG +\
-				  HAL_REO1_QDESC_LUT_BASE0(ab);
-			ath12k_hif_write32(ab, address, 0);
-		}
-
-		if (dp->ml_reoq_lut.vaddr) {
-			address = HAL_SEQ_WCSS_UMAC_REO_REG +\
-				  HAL_REO1_QDESC_LUT_BASE1(ab);
-			ath12k_hif_write32(ab, address, 0);
+		if (test_bit(ATH12K_FLAG_QMI_FW_READY_COMPLETE, &ab->dev_flags)){
+			if (dp->reoq_lut.vaddr) {
+				address = HAL_SEQ_WCSS_UMAC_REO_REG +\
+					  HAL_REO1_QDESC_LUT_BASE0(ab);
+				ath12k_hif_write32(ab, address, 0);
+			}
+
+			if (dp->ml_reoq_lut.vaddr) {
+				address = HAL_SEQ_WCSS_UMAC_REO_REG +\
+					  HAL_REO1_QDESC_LUT_BASE1(ab);
+				ath12k_hif_write32(ab, address, 0);
+			}
 		}
-
 		ath12k_hif_power_down(ab);
 
 		mutex_unlock(&ab->core_lock);
-- 
2.34.1

